android/src/main/java/io/radar/react/RNRadarModule.java

```diff
@@ -109,29 +163,69 @@ public void updateLocation(ReadableMap locationMap, final Promise promise) {
-                if (promise == null)
+            public void onComplete(@NonNull Radar.RadarStatus status, @Nullable Location location, @Nullable RadarEvent[] events, @Nullable RadarUser user) {
+                if (promise == null) {
                     return;
+                }
 
-                if (status == Radar.RadarStatus.SUCCESS) {
-                    WritableMap map = Arguments.createMap();
-                    map.putString("status", RNRadarUtils.stringForStatus(status));
-                    if (location != null)
-                        map.putMap("location", RNRadarUtils.mapForLocation(location));
-                    if (events != null)
-                        map.putArray("events", RNRadarUtils.arrayForEvents(events));
-                    if (user != null)
-                        map.putMap("user", RNRadarUtils.mapForUser(user));
-                    promise.resolve(map);
-                } else {
-                    promise.reject(RNRadarUtils.stringForStatus(status), RNRadarUtils.stringForStatus(status));
+                try {
+                    if (status == Radar.RadarStatus.SUCCESS) {
+                        WritableMap map = Arguments.createMap();
+                        map.putString("status", status.toString());
+                        if (location != null) {
+                            map.putMap("location", RNRadarUtils.mapForJson(Radar.jsonForLocation(location)));
+                        }
+                        if (events != null) {
+                            map.putArray("events", RNRadarUtils.arrayForJson(RadarEvent.toJson(events)));
+                        }
+                        if (user != null) {
+                            map.putMap("user", RNRadarUtils.mapForJson(user.toJson()));
+                        }
+                        promise.resolve(map);
+                    } else {
+                        promise.reject(status.toString(), status.toString());
+                    }
+                } catch (JSONException e) {
+                    promise.reject(Radar.RadarStatus.ERROR_SERVER.toString(), Radar.RadarStatus.ERROR_SERVER.toString());
                 }
             }
         });
     }
 
+    @ReactMethod
+    public void startTrackingEfficient() {
+        Radar.startTracking(RadarTrackingOptions.EFFICIENT);
+    }
+
+    @ReactMethod
+    public void startTrackingResponsive() {
+        Radar.startTracking(RadarTrackingOptions.RESPONSIVE);
+    }
+
+    @ReactMethod
+    public void startTrackingContinuous() {
+        Radar.startTracking(RadarTrackingOptions.CONTINUOUS);
+    }
+
+    @ReactMethod
+    public void startTrackingCustom(ReadableMap optionsMap) {
+        try {
+            JSONObject optionsObj = RNRadarUtils.jsonForMap(optionsMap);
+            RadarTrackingOptions options = RadarTrackingOptions.fromJson(optionsObj);
+            Radar.startTracking(options);
+        } catch (JSONException e) {
```

Comment:
is it possible for `json` serialization of an `optionsMap` to error out?

Reply:
Yeah, the `optionsMap` methods can throw `JSONException`. (I'm not sure why React Native uses these instead of `JSONObject`s in the first place...)

Reply:
yeah, definitely odd, let's bubble up this exception to the caller then?